### PR TITLE
Adding sharedid submodule

### PR DESCRIPTION
**Pull request by SKOCHERI**
## Type of change

- [ ] Bugfix
- [ ] Feature
- [ ] New bidder adapter   
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Does this change affect user-facing APIs or examples documented on http://prebid.org?
- [ ] Other

## Description of change

- test parameters for validating bids
```
{
  bidder: '<bidder name>',
  params: {
    // ...
  }
}
```

Be sure to test the integration with your adserver using the [Hello World](/integrationExamples/gpt/hello_world.html) sample page.

- contact email of the adapter’s maintainer
- [ ] official adapter submission

For any changes that affect user-facing APIs or example code documented on http://prebid.org, please provide:

- A link to a PR on the docs repo at https://github.com/prebid/prebid.github.io/

### CODE REVIEW

File: integrationExamples/gpt/userId_example.html [hunk 1]
```diff
@@ -184,9 +184,16 @@
                         name: 'idl_env',
                         expires: 30
                       }
+                    }, {
+                      name: "sharedId",
+                      storage: {
+                        type: "cookie",
+                        name: "sharedId",
```
Comment (brodrigu) [inline integrationExamples/gpt/userId_example.html]:
```suggestion
                        name: "sharedid",
```
it looks like storage.name is lowercase in all other examples, we should probably keep that consistency.

File: modules/sharedIdSystem.js [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+   */
+  name: MODULE_NAME,
+
+  sharedIdFactory: undefined,
+
+  /**
+   * decode the stored id value for passing to bid requests
+   * @function
+   * @param {string} value
+   * @returns {{sharedId:string}}
```
Comment (brodrigu) [inline modules/sharedIdSystem.js]: this could also return `undefined`. if intentional, we should indicate here.

File: modules/sharedIdSystem.js [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+  /**
+   * performs action to obtain id and return a value.
+   * @function
+   * @param {SubmoduleParams} [configParams]
+   * @returns {sharedId}
+   */
+  getId: function(configParams) {
+    let sharedId = this.sharedIdGenerator();
+    return {
+      id: sharedId }
```
Comment (brodrigu) [inline modules/sharedIdSystem.js]:
```suggestion
      id: sharedId
    }
```

File: modules/sharedIdSystem.js [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+  },
+
+  sharedIdGenerator: function () {
```
Comment (brodrigu) [inline modules/sharedIdSystem.js:50]: ideally, we should add docBlocks for all functions.

File: modules/sharedIdSystem.js [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    if (!this.sharedIdFactory) {
+      this.sharedIdFactory = this.factory();
+    }
+
+    return this.sharedIdFactory();
+  },
+
+  factory: function (currPrng) {
```
Comment (brodrigu) [inline modules/sharedIdSystem.js:58]: I had to google what `Prng` stands for, would be good to indicate what that is here. A docBlock would solve for this.

File: modules/sharedIdSystem.js [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    if (!currPrng) {
+      currPrng = this.detectPrng();
+    }
+    return function ulid(seedTime) {
+      if (isNaN(seedTime)) {
+        seedTime = Date.now();
+      }
+      return this.encodeTime(seedTime, TIME_LEN) + this.encodeRandom(RANDOM_LEN, currPrng);
+    };
```
Comment (brodrigu) [inline modules/sharedIdSystem.js:78]: in `sharedIdGenerator` we pull out `sharedIdFactory` to its own, independent, class method. Here we are defining `ulid` inline. Is this necessary? if not, I might suggest pulling it into a separate method and reference it here instead.

File: modules/sharedIdSystem.js [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    if (isNaN(now)) {
+      throw new Error(now + ' must be a number');
+    }
+    if (now > TIME_MAX) {
+      throw this.createError('cannot encode time greater than ' + TIME_MAX);
+    }
+    if (now < 0) {
+      throw this.createError('time must be positive');
+    }
+    if (Number.isInteger(now) === false) {
```
Comment (brodrigu) [inline modules/sharedIdSystem.js]: you should check this condition before comparing to `0` or `TIME_MAX`

Comment (brodrigu) [reply modules/sharedIdSystem.js]: actually just realized there is a `isNaN` first, so maybe not necessary.

Comment (brodrigu) [reply modules/sharedIdSystem.js]: actually, `isNaN('1')` is `false` and `Number.isInteger('1')` is also false, so though I think this more or less safe, still better to clear assumptions first.

File: modules/sharedIdSystem.js [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+      throw this.createError('time must be an integer');
+    }
+    let mod;
+    let str = '';
+    for (; len > 0; len--) {
```
Comment (brodrigu) [inline modules/sharedIdSystem.js:97]: do we want to error out if `len` is not a number? or is returning an empty string fine?

File: modules/sharedIdSystem.md [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+### Prebid Params
+
+Individual params may be set for the Shared ID User ID Submodule. 
+```
+pbjs.setConfig({
+    usersync: {
+        userIds: [{
+            name: 'sharedId',
+            storage: {
+                name: 'sharedId',
```
Comment (brodrigu) [inline modules/sharedIdSystem.md]:
```suggestion
                name: 'sharedid',
```

File: modules/sharedIdSystem.md [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+### Parameter Descriptions for the `usersync` Configuration Section
+The below parameters apply only to the Shared ID User ID Module integration.
+
+| Params under usersync.userIds[]| Scope | Type | Description | Example |
+| --- | --- | --- | --- | --- |
+| name | Required | String | ID value for the Shared ID module - `"sharedId"` | `"sharedId"` |
+| storage | Required | Object | The publisher must specify the local storage in which to store the results of the call to get the user ID. This can be either cookie or HTML5 storage. | |
+| storage.type | Required | String | This is where the results of the user ID will be stored. The recommended method is `localStorage` by specifying `html5`. | `"html5"` |
+| storage.name | Required | String | The name of the cookie or html5 local storage where the user ID will be stored. | `"sharedId"` |
```
Comment (brodrigu) [inline modules/sharedIdSystem.md]:
```suggestion
| storage.name | Required | String | The name of the cookie or html5 local storage where the user ID will be stored. | `"sharedid"` |
```

File: modules/userId/eids.js [hunk 11]
```diff
@@ -81,6 +81,11 @@ const USER_IDS_CONFIG = {
   'netId': {
     source: 'netid.de',
     atype: 1
+  },
+  // sharedId
+  'sharedId': {
+    source: 'rubiconproject.com',
```
Comment (brodrigu) [inline modules/userId/eids.js]:
```suggestion
    source: 'sharedid.org',
```
Need to confirm with Garrett, but I believe we don't want rubicon as the source.

File: modules/userId/eids.md [hunk 12]
```diff
@@ -85,6 +85,13 @@ userIdAsEids = [
             id: 'some-random-id-value',
             atype: 1
         }]
+    },
+    {
+        source: 'rubiconproject.com',
```
Comment (brodrigu) [inline modules/userId/eids.md]:
```suggestion
        source: 'sharedid.org',
```
again, pending confirmation with Garrett

File: modules/userId/userId.md [hunk 13]
```diff
@@ -64,6 +64,13 @@ pbjs.setConfig({
                 name: '_li_pbid',
                 expires: 60
             }
+        }, {
+             name: 'sharedId',
+             storage: {
+                 type: 'cookie',
+                 name: 'sharedId',
```
Comment (brodrigu) [inline modules/userId/userId.md]:
```suggestion
                 name: 'sharedid',
```

File: modules/userId/userId.md [hunk 14]
```diff
@@ -113,6 +120,13 @@ pbjs.setConfig({
                  name: '_li_pbid',
                  expires: 60
              }
+        }, {
+             name: 'sharedId',
+             storage: {
+                type: 'cookie',
+                name: 'sharedId',
```
Comment (brodrigu) [inline modules/userId/userId.md]:
```suggestion
                name: 'sharedid',
```

File: test/spec/modules/eids_spec.js [hunk 15]
```diff
@@ -146,6 +146,17 @@ describe('eids array generation for known sub-modules', function() {
...
     });
   });
+  it('SharedId', function() {
+    const userId = {
+      sharedId: 'test_sharedId'
+    };
+    const newEids = createEidsArray(userId);
+    expect(newEids.length).to.equal(1);
+    expect(newEids[0]).to.deep.equal({
+      source: 'rubiconproject.com',
```
Comment (brodrigu) [inline test/spec/modules/eids_spec.js]:
```suggestion
      source: 'sharedid.org',
```

File: test/spec/modules/userId_spec.js [hunk 16]
```diff
@@ -359,14 +361,17 @@ describe('User ID', function() {
           }, {
             name: 'netId',
             storage: { name: 'netId', type: 'cookie' }
-          }]
+          }, {
+            name: 'sharedId',
+            storage: { name: 'sharedId', type: 'cookie' }
```
Comment (brodrigu) [inline test/spec/modules/userId_spec.js]:
```suggestion
            storage: { name: 'sharedid', type: 'cookie' }
```

File: test/spec/modules/userId_spec.js [hunk 17]
```diff
@@ -786,6 +791,52 @@ describe('User ID', function() {
       }, {adUnits});
     });
 
+    it('test hook from sharedId html5', function(done) {
+      // simulate existing browser local storage values
+      localStorage.setItem('sharedId', JSON.stringify({'sharedId': 'test_sharedId'}));
```
Comment (brodrigu) [inline test/spec/modules/userId_spec.js]: I'm not 100% sure where in this line (and elsewhere in this file) `sharedId` would need to be changed to `sharedid`

**Review by brodrigu — CHANGES_REQUESTED** [review]

**Comment by jdwieland8282** [comment]
Docs PR: https://github.com/prebid/prebid.github.io/pull/1923

File: modules/sharedIdSystem.js [hunk 18]
```diff
@@ -0,0 +1,176 @@
...
+    if (!this.sharedIdFactory) {
+      this.sharedIdFactory = this.factory();
+    }
+
+    return this.sharedIdFactory();
+  },
+
+  /**
+   * the factory to generate unique identifier based on time and current pseudorandom number
+   * @param {string} the current pseudorandom number
```
Comment (brodrigu) [inline modules/sharedIdSystem.js]:
```suggestion
   * @param {string} the current pseudorandom number generator
```

File: modules/sharedIdSystem.js [hunk 19]
```diff
@@ -0,0 +1,176 @@
...
+    let sharedId = this.sharedIdGenerator();
+    return {
+      id: sharedId
+    }
+  },
+  /**
+   * the shared id generator factory.
+   * @returns {*}
+   */
+  sharedIdGenerator: function () {
```
Comment (yava-ua) [inline modules/sharedIdSystem.js]: this and below methods do not seem to be related to the submodule. It seems safe to extract them out of the module and just reuse it here. Even if we want to leave factory for submodule use, other methods may be still regular functions

**Comment by SKOCHERI** [comment]
There has been changes with the sharedId generation hence closing this pull request

**Pull request closed without merging**